### ISSUE TITLE
Incorrect label shadow warning

### ISSUE THREAD

**Issue by WiSaGaN** [rust-lang/rust#88225]
I tried this code:

```rust
fn main() {
    if true {
        'l : loop {
            break 'l;
        }
    } else {
        'l : loop {
            break 'l;
        }
    }
}
```

I expected to see this happen: *no warning regarding to label*

Instead, this happened: *got warning about label shadowing*

```
warning: label name `'l` shadows a label name that is already in scope
 --> src/main.rs:7:9
  |
3 |         'l : loop {
  |         -- first declared here
...
7 |         'l : loop {
  |         ^^ label `'l` already in scope
```

version
```
Build using the Nightly version: 1.56.0-nightly

(2021-08-21 d3e2578c31688619ddc0)
```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9b6bec99ce1b325bcbf04e2b511ffb14)

**Comment by ehuss** [comment]
Thanks for the report!  This is currently intentional, and a know issue.  The warning could be clearer, though. Closing as a duplicate of #24278 and #31745.

**Comment by WiSaGaN** [comment]
Ok. I digged up an exact duplicate here: https://github.com/rust-lang/rust/issues/81799

This warning is very unintuitive. @ehuss can you provide a link to the reason why this is intended?

**Comment by ehuss** [comment]
I don't think there is any information beyond what is linked in those issues. This was something added before 1.0, and it hasn't been revisited since.  The *why* is primarily explained in https://internals.rust-lang.org/t/psa-rejecting-duplicate-loop-labels/1833, where there was an intent to add explicit syntax for labeling a region for a borrow, but that never got beyond the idea stage (AFAIK).  #24278 also explains how there are macro hygiene problems.
